lib/nm/index.ts

```diff
@@ -0,0 +1,121 @@
+	wifiDevice: WifiDevice;
+	accessPoints: AccessPoint[];
+	wifiActive: boolean;
+
+	constructor() {
+		super();
+
+		this.wifiActive = true;
+
+		this.getWifiDevice()
+		.then((device: WifiDevice) => {
+			const service = Promise.promisifyAll(systemBus.getService('org.freedesktop.NetworkManager'));
+			service.getInterfaceAsync(device.devicePath, 'org.freedesktop.NetworkManager.Device.Wireless')
+			.then((iface) => {
+				console.log('YOLO');
+				iface.on('AccessPointAdded', (...data) => {
+					console.log('YAY ACCESS POINT ADDED', ...data);
+				});
+				iface.on('AccessPointRemoved', (...data) => {
+					console.log('YAY ACCESS POINT REMOVED', ...data);
+				});
+				console.log('SBOLO');
+			})
+			.catch((err) => {
+				console.error('gnappo', err);
+			});
+		});
+	}
+
+	toggleWifi(value) {
+		return this.wifiActive = value;
+	}
+
+	getBus() {
+		return systemBus;
+	}
+
+	callMethod(path: string[]|string = []) {
+		return (iface: string[]|string = []) => ([method, signature]) => (params: any = []) => {
+			return new Promise((resolve, reject) => {
+				// {
+				//   path: `/org/freedesktop/NetworkManager/Devices/1`, // Object fs path
+				//   destination: service.join('.'), // ['org', 'freedesktop', 'NetworkManager'] // Object
+				//   iface: 'org.freedesktop.DBus.Properties', // Object interface
+				//   member: 'Get', // interface method
+				//   signature: 'ss', // interface method params types
+				//   body: ['org.freedesktop.NetworkManager.Device', 'DeviceType'], // [<interface to get property of>, <property name>] // params
+				// }
+				const command = {
+					path: _.isString(path) ? path : `/${service.concat(path).join('/')}`,
+					destination: service.join('.'),
+					interface: _.isString(iface) ? iface : service.concat(iface).join('.'),
+					member: method,
+					signature,
+					body: params,
+				};
+				console.log('command', command.path, command.interface, method, ...params);
+				systemBus.invoke(command,
+				function(err, ...results) {
+					if (err) {
+						return reject(err);
+					}
+					return resolve(...results);
+				});
+			});
+		};
+	}
+
+	getWifiDevice() {
+		return this.callMethod([])([])(['GetDevices', ''])([])
+		.then((devices: string[]) => {
+			const getDevicesProperty = (prop: string[]) => _.map(devices, (devicePath) => {
+				return new Promise((resolve) => {
+					this.callMethod(devicePath)('org.freedesktop.DBus.Properties')(['Get', 'ss'])(prop)
+					.then((type) => {
+						const [key, [value]]: any = type;
+						return resolve({
+							devicePath,
+							type: value,
+						});
+					});
+				});
+			});
+			return Promise.all(getDevicesProperty(['org.freedesktop.NetworkManager.Device', 'DeviceType']))
+			.then((devices: WifiDevice[]) => {
+				const wifiDevices: any = _.filter(devices, ({type}) => {
+					return type === NetworkManager.DEVICE_TYPE.WIFI;
+				});
+				return this.wifiDevice = wifiDevices[0];
+			})
+			.catch((err) => {
+				return err;
```

Comment:
This does not make sense to me. This function should return a device or throw an error, not return an error.

Reply:
To my knowledge, `Promise` `catch` follows the same rules as `then`, so an error returned inside a catch statement should also be returned to each and every catch defined in the defined Promise flow...
Which is also why I was able to receive errors, I guess

Reply:
What I mean is that if you catch an error that you want to rethrow, you'd write

```typescript
.catch((error) => {
    throw error;
})
```
which is the same as writing nothing.

If you `return` an error in a catch, you may get it as a value in the next `then`:

```typescript
asyncFunctionThatMayThrow()
.catch((error) => {
    return error;
})
.then((value) => {
    // value may be an Error here
});
```

Reply:
Uhm, strange, I was able to receive errors even though I used `return` instead of `throw`.
Anyways, changed it accordingly to avoid any doubt.